### PR TITLE
New functionality to allow merge upsert on existing table.

### PR DESCRIPTION
**Pull request by jiteshsoni**
## Functionality: This will allow us to update else insert (merge upsert) on the primary key for an existing Glue table

This resolves issue https://github.com/awslabs/aws-data-wrangler/issues/503 #510

### CODE REVIEW

**Review by igorborgest — CHANGES_REQUESTED** [review]
Thanks @jiteshsoni !

The implementation seems great!
Could you also add tests using the `merge_upsert_table()`, please?

**Comment by jiteshsoni** [comment]
> merge_upsert_table()

@igorborgest Testing that function would require some kind of aws account connectivity. Could you suggest me how to do that

**Comment by igorborgest** [comment]
Take a look if [this guide](https://github.com/awslabs/aws-data-wrangler/blob/master/CONTRIBUTING.md#data-lake-test-environment) can help you, please.

**Comment by jiteshsoni** [comment]
> Take a look if [this guide](https://github.com/awslabs/aws-data-wrangler/blob/master/CONTRIBUTING.md#data-lake-test-environment) can help you, please.

Thank you for your response. I will try it out.

**Comment by jiteshsoni** [comment]
Made changes to address https://github.com/awslabs/aws-data-wrangler/pull/510#pullrequestreview-564216758
https://github.com/awslabs/aws-data-wrangler/pull/510/files

File: awswrangler/s3/_merge_upsert_table.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
+"""Amazon Merge Upsert Module which can upsert existing table (PRIVATE)."""
+
+import logging
+from typing import List
+
+import pandas
+
+import awswrangler as wr
+
+logging.basicConfig(level=logging.INFO)
```
Comment (igorborgest) [inline awswrangler/s3/_merge_upsert_table.py:10]: You can remove the first three lines here. Only the `_logger: logging.Logger = logging.getLogger(__name__)` is necessary.

File: awswrangler/s3/_merge_upsert_table.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    if wr.catalog.does_table_exist(database=database, table=table):
+        # Read the existing table into a pandas dataframe
+        existing_df = wr.s3.read_parquet_table(database=database, table=table)
+        # Check if data quality inside dataframes to be merged are sufficient
+        if _is_data_quality_sufficient(existing_df=existing_df, delta_df=delta_df, primary_key=primary_key) is True:
+            # If data quality is sufficient then merge upsert the table
+            _update_existing_table(
+                existing_df=existing_df, delta_df=delta_df, primary_key=primary_key, database=database, table=table
+            )
+    elif wr.catalog.does_table_exist(database=database, table=table) is False:
```
Comment (igorborgest) [inline awswrangler/s3/_merge_upsert_table.py:70]: This `elif` does not make sense here. Why not just a regular `else`?

File: awswrangler/s3/_merge_upsert_table.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+            f"Column names or number of columns mismatch! \n Columns in delta_df {delta_df.columns}.\n  Columns in "
+            f"existing_df is {existing_df.columns} "
+        )
+    # Return True only if no errors are encountered
+    _logger.info("error_messages %s", str(error_messages))
+    return len(error_messages) == 0
+
+
+def merge_upsert_table(delta_df: pandas.DataFrame, database: str, table: str, primary_key: List[str]) -> None:
+    """Perform Upsert (Update else Insert) onto an existing Glue table."""
```
Comment (igorborgest) [inline awswrangler/s3/_merge_upsert_table.py:59]: This is a public function so we need to have a proper docstring with arguments descriptions and examples like we have in the others modules.

File: awswrangler/s3/_merge_upsert_table.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        error_messages.append(
+            f"Column names or number of columns mismatch! \n Columns in delta_df {delta_df.columns}.\n  Columns in "
+            f"existing_df is {existing_df.columns} "
+        )
+    # Return True only if no errors are encountered
+    _logger.info("error_messages %s", str(error_messages))
+    return len(error_messages) == 0
+
+
+def merge_upsert_table(delta_df: pandas.DataFrame, database: str, table: str, primary_key: List[str]) -> None:
```
Comment (igorborgest) [inline awswrangler/s3/_merge_upsert_table.py:58]: I think you should consider others optional arguments to be passed forward to the inner function calls. At least `boto3_session` is a must have.

File: awswrangler/s3/_merge_upsert_table.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    if (
+        existing_df.shape[1] != delta_df.shape[1]
+        or len(existing_df.columns.intersection(delta_df.columns)) != existing_df.shape[1]
+    ):
+        error_messages.append(
+            f"Column names or number of columns mismatch! \n Columns in delta_df {delta_df.columns}.\n  Columns in "
+            f"existing_df is {existing_df.columns} "
+        )
+    # Return True only if no errors are encountered
+    _logger.info("error_messages %s", str(error_messages))
```
Comment (igorborgest) [inline awswrangler/s3/_merge_upsert_table.py:54]: Isn't this string cast redundant?

File: tests/test_s3_merge_upsert.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    # Fail as table does not exist
+    delta_df = pd.DataFrame([[1, 2, 3], [4, 5, 6], [7, 8, 9]], columns=["col_a", "col_b", "col_c"])
+    primary_key = ["col_a", "col_b"]
+    with pytest.raises(AttributeError):
+        merge_upsert_table(
+            delta_df=delta_df, database="xxx_does_not_exist", table="xxx_does_not_exist", primary_key=primary_key
+        )
+
+
+# pytest tests/test_s3_merge_upsert.py::test_success_case
```
Comment (igorborgest) [inline tests/test_s3_merge_upsert.py:66]: Remove this comment please.

**Review by igorborgest — CHANGES_REQUESTED** [review]
We are in the right way!

File: awswrangler/s3/_merge_upsert_table.py (hunk 1 above)
Comment (jiteshsoni) [reply awswrangler/s3/_merge_upsert_table.py:10]: Done

File: awswrangler/s3/_merge_upsert_table.py (hunk 2 above)
Comment (jiteshsoni) [reply awswrangler/s3/_merge_upsert_table.py:70]: Makes sense. I have simplified it.

File: awswrangler/s3/_merge_upsert_table.py (hunk 5 above)
Comment (jiteshsoni) [reply awswrangler/s3/_merge_upsert_table.py:54]: error_messages is a list data type. This we need to convert the list to str before we write it out.

File: tests/test_s3_merge_upsert.py (hunk 6 above)
Comment (jiteshsoni) [reply tests/test_s3_merge_upsert.py:66]: Done

File: awswrangler/s3/_merge_upsert_table.py (hunk 3 above)
Comment (jiteshsoni) [reply awswrangler/s3/_merge_upsert_table.py]: Added doc string and examples

File: awswrangler/s3/_merge_upsert_table.py (hunk 4 above)
Comment (jiteshsoni) [reply awswrangler/s3/_merge_upsert_table.py]: Added boto3_session argument

**Comment by jiteshsoni** [comment]
@igorborgest I have made the changes you had suggested. Please let me know if you see any other issue. 


Screen Shot 2021-01-15 at 01 55 42

File: awswrangler/s3/_merge_upsert_table.py (hunk 5 above)
Comment (igorborgest) [reply awswrangler/s3/_merge_upsert_table.py:64]: I think we don't need that.

image

File: tests/test_s3_merge_upsert.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+def test_is_data_quality_sufficient_fail_for_duplicate_data():
+    # Check for duplicate data inside the dataframe
+    existing_df = pd.DataFrame([[1, 2, 3], [1, 2, 3], [7, 8, 9], [10, 11, 12]], columns=["col_a", "col_b", "col_c"])
+    delta_df = pd.DataFrame([[1, 2, 3], [4, 5, 6], [7, 8, 9]], columns=["col_a", "col_b", "col_c"])
+    primary_key = ["col_a", "col_b"]
+    assert _is_data_quality_sufficient(existing_df=existing_df, delta_df=delta_df, primary_key=primary_key) is False
+
+
+# pytest tests/test_s3_merge_upsert.py::test_merge_upsert_table_fail_as_table_does_not_exist
```
Comment (igorborgest) [inline tests/test_s3_merge_upsert.py:55]: I think you forgot to clean up this comment too.

**Review by igorborgest — CHANGES_REQUESTED** [review]
Thanks for all changes. There are only a couple now.

**Comment by jiteshsoni** [comment]
> Thanks for all changes. There are only a couple now.

@igorborgest I have incorporated your suggestions.

File: awswrangler/s3/_merge_upsert_table.py [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+    """Perform Update else Insert onto an existing Glue table."""
+    # Set Index on the pandas dataframe so that join/concat can be made
+    existing_df = existing_df.set_index(keys=primary_key, drop=False, verify_integrity=True)
+    delta_df = delta_df.set_index(keys=primary_key, drop=False, verify_integrity=True)
+    # Merge-Upsert the data for both of the dataframe
+    merged_df = pandas.concat([existing_df[~existing_df.index.isin(delta_df.index)], delta_df])
+    # Remove the index and drop the index columns
+    merged_df = merged_df.reset_index(drop=True)
+    # Get existing tables location
+    path = wr.catalog.get_table_location(database=database, table=table)
```
Comment (igorborgest) [inline awswrangler/s3/_merge_upsert_table.py:31]: Oh you are missing to pass the `boto3_session` argument forward here in the `get_table_location()` call.

File: awswrangler/s3/_merge_upsert_table.py [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+    --------
+    Reading all Parquet files under a prefix
+    >>> import awswrangler as wr
+    >>> import pandas as pd
+    >>> delta_df = pd.DataFrame({"id": [1], "cchar": ["foo"], "date": [datetime.date(2021, 1, 2)]})
+    >>> primary_key = ["id", "cchar"]
+    >>> wr.s3.merge_upsert_table(delta_df=delta_df, database='database', table='table', primary_key=primary_key)
+    """
+    # Check if table exists first
+    if wr.catalog.does_table_exist(database=database, table=table):
```
Comment (igorborgest) [inline awswrangler/s3/_merge_upsert_table.py:105]: It is missing the boto3_session argument here too.

File: awswrangler/s3/_merge_upsert_table.py [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+        # Read the existing table into a pandas dataframe
+        existing_df = wr.s3.read_parquet_table(database=database, table=table)
```
Comment (igorborgest) [inline awswrangler/s3/_merge_upsert_table.py:107]: It is missing the boto3_session argument here too.

**Review by igorborgest — CHANGES_REQUESTED** [review]
I've just noticed you are not propagating the `boto3_session` correctly in all Wrangler function call.

**Review by igorborgest — APPROVED** [review]
LGTM! Thanks @jiteshsoni !

**Pull request merged**